### PR TITLE
Detect more cases of an entry modified in the external changes resolver dialog

### PR DESCRIPTION
**Pull request by HoussemNasri**
**Context**
When an external change in the database occurs, JabRef loads the new version of the database and parses it. Then it compares the list of entries on JabRef and the list of entries loaded from the new version. In the comparison logic, JabRef tries to predict what entries were added, deleted, or modified. The logic is straightforward:
- If an entry exists in JabRef's memory but not in the external database then it's **deleted** from outside
- If an entry exists in the external database but not in JabRef's memory then it's **added** from outside
- If an entry exists in both places then it is **modified** from outside

**So how does JabRef know when an entry is modified?**
JabRef doesn't know for sure when an entry is modified but it can make some decent predictions. To decide if "entry A" was modified JabRef compares "entry A" against all entries in the external database and tries to find the most similar entry. After finding the most similar entry, it compares the similarity score against a threshold, and only when the similarity score is above that threshold then it can predict that "entry A" was modified.

**Why this PR?**
The aim of this pr is to detect more cases of an entry modified that the current similarity detection algorithm can't detect. The pr addresses two main cases:
- When you add or modify many fields of an entry without altering the citation key value.
- When you add or modify many fields of an entry (Citation key included) but the entry is still identifiable using other fields *(ISBN, DOI, etc)*.
____
The issue was first reported in https://github.com/JabRef/jabref-issue-melting-pot/issues/18
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java [hunk 1]
```diff
@@ -83,28 +69,26 @@ private static List<BibEntryDiff> compareEntries(List<BibEntry> originalEntries,
         }
 
         // Now we've found all exact matches, look through the remaining entries, looking for close matches.
-        for (Iterator<BibEntry> iteratorNotMatched = notMatched.iterator(); iteratorNotMatched.hasNext(); ) {
-            BibEntry originalEntry = iteratorNotMatched.next();
-
+        DuplicateCheck duplicateCheck = new DuplicateCheck(Globals.entryTypesManager);
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java:72]: This wont work. The EntryTypesManager needs to be injected in the method arguments. And Globals is a deprecated GUI only class.

File: src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java [hunk 2]
```diff
@@ -83,28 +69,26 @@ private static List<BibEntryDiff> compareEntries(List<BibEntry> originalEntries,
         }
 
         // Now we've found all exact matches, look through the remaining entries, looking for close matches.
-        for (Iterator<BibEntry> iteratorNotMatched = notMatched.iterator(); iteratorNotMatched.hasNext(); ) {
-            BibEntry originalEntry = iteratorNotMatched.next();
-
+        DuplicateCheck duplicateCheck = new DuplicateCheck(new BibEntryTypesManager());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java:72]: The BibEntryTypesManager contains sets with all entry types and their fields. I really would avoid creating this class every time. Ideally you pass the reference down or at least make it an instance variable

**Review by Siedlerchr — APPROVED** [review]
just a small suggestion, otherwise lgtm

File: src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java (hunk 2 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java:72]: Decided to do this in a Follow up pr

**Review by Siedlerchr — APPROVED** [review]

**Comment by calixtus** [comment]
BibEntryTypesManager will be addressed in a follow up pr. Should probably be a static class or an injected value.

**Comment by koppor** [comment]
> The pr addresses two main cases:
> 
>     * When you add or modify many fields of an entry without altering the citation key value.
>     * When you add or modify many fields of an entry (Citation key included) but the entry is still identifiable using other fields _(ISBN, DOI, etc)_.

Thank you so much for working on this! - I always had this in mind. Think, this is a huge step forward in handling complex libraries (duplicate finding, ...).

Would it be possible to add some of your text to the class of BibDatabaseDiff.java?

**Comment by HoussemNasri** [comment]
Sure, I will create a pr soon. I should also hoist the BibEntryTypesManager dependency, almost forgot about that.

**Comment by calixtus** [comment]
Careful with the bibentrytypesmanager, I'm currently working on a complete refactor of this whole complex to fit to the new prefs scheme

**Pull request merged**